examples/agent_client/examples/agent_client.rs

```diff
@@ -0,0 +1,324 @@
+        log_level: RDKafkaLogLevel::Info,
+        talos_type: TALOS_TYPE,
+    };
+
+    (cfg_agent, cfg_kafka)
+}
+
+fn make_candidate(xid: String) -> CertificationRequest {
+    let tx_data = CandidateData {
+        xid,
+        readset: Vec::new(),
+        readvers: Vec::new(),
+        snapshot: 5,
+        writeset: Vec::from(["3".to_string()]),
+    };
+
+    CertificationRequest {
+        message_key: "12345".to_string(),
+        candidate: tx_data,
+    }
+}
+
+fn name_rate(v: Duration) -> String {
+    let nr = 1_f64 / v.as_secs_f64();
+    if nr > 1000_f64 {
+        format!("{}k/s", nr / 1000_f64)
+    } else {
+        format!("{}/s", nr)
+    }
+}
+
+// async fn make_agent() -> Box<TalosAgentType> {
+//     let (cfg_agent, cfg_kafka) = make_configs();
+//
+//     TalosAgentBuilder::new(cfg_agent)
+//         .with_kafka(cfg_kafka)
+//         .build()
+//         .unwrap_or_else(|e| panic!("{}", format!("Unable to build agent {}", e)))
+// }
+
+async fn make_agentv2(publish_times: &Arc<Mutex<HashMap<String, u64>>>) -> Box<TalosAgentType> {
+    let (cfg_agent, cfg_kafka) = make_configs();
+
+    TalosAgentBuilder::new(cfg_agent)
+        .with_kafka(cfg_kafka)
+        .build_v2(publish_times)
+        .await
+        .unwrap_or_else(|e| panic!("{}", format!("Unable to build agent {}", e)))
+}
+
+#[tokio::main]
+async fn main() -> Result<(), String> {
+    env_logger::init();
+
+    thread::sleep(Duration::from_secs(5));
+
+    certify(BATCH_SIZE).await
+}
+
+async fn certify(batch_size: i32) -> Result<(), String> {
+    info!("Certifying {} transactions", batch_size);
+
+    // "global" state where we will collect publishing times: A - B, where
+    // B: transaction start time
+    // A: the end of call to kafka send_message(candidate)
+    let publish_times: Arc<Mutex<HashMap<String, u64>>> = Arc::new(Mutex::new(HashMap::new()));
+    let mut tasks = Vec::<JoinHandle<Result<Timing, String>>>::new();
+    let agent = Arc::new(make_agentv2(&publish_times).await);
+
+    // Allow some time for consumer to properly connect
+    thread::sleep(Duration::from_secs(5));
+
+    let started_at = OffsetDateTime::now_utc().unix_timestamp_nanos();
+    info!("Starting publishing process...");
+
+    let adjust_every = 100;
+    let mut delay = BASE_DELAY;
+    let mut done = 0;
+    let mut skip = 0;
+    let min_sleep = Duration::from_secs_f32(0.0001);
+    for i in 0..batch_size {
+        done += 1;
+        if skip == 0 {
+            thread::sleep(delay);
+        } else {
+            skip -= 1
+        }
+        let ac = Arc::clone(&agent);
+        let task = tokio::spawn(async move { Timing::capture(|| async { ac.certify(make_candidate(Uuid::new_v4().to_string())).await }).await });
+        tasks.push(task);
+
+        let p = i + 1;
+        if p % PROGRESS_EVERY == 0 || p == batch_size {
+            info!("Published {} of {}", p, batch_size);
+        }
+
+        if done % adjust_every == 0 {
```

Comment:
Little confused on this one, shouldn't the delay be done where the task is consumed and awaited? 

Because when you introduce the delay here, it just delays the rate at which the task is added to the Vec, but not the rate at which it is executed.

i.e the ` for task in tasks ` loop?

Reply:
Yep that is the intention. Delay before spawning a new task so that our intended rate is matching real rate. No need to wait for task to finish before spawning a new one. Otherwise publishing will become sequential, right?

Reply:
when you spawn a task and attach it to a handle, the task is executed only when you  `await` on the handle. i.e in the above code, the publish actually happens inside the for loop of `tasks`. Therefore it is actually sequential at the moment.

Reply:
Sorry you are right, I just got mixed up with something else. Ignore this comment :)